Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,775 @@
+instance instProdMetricSpace [MetricSpace α] [MetricSpace β] : MetricSpace (WithLp p (α × β)) :=
+  MetricSpace.ofT0PseudoMetricSpace _
+
+variable {p α β}
+
+theorem prod_nndist_eq_add [PseudoMetricSpace α] [PseudoMetricSpace β]
+    (hp : p ≠ ∞) (x y : WithLp p (α × β)) :
+    nndist x y = (nndist x.fst y.fst ^ p.toReal + nndist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) :=
+  NNReal.eq <| by
+    push_cast
+    exact prod_dist_eq_add (p.toReal_pos_iff_ne_top.mpr hp) _ _
+
+theorem prod_nndist_eq_sup [PseudoMetricSpace α] [PseudoMetricSpace β] (x y : WithLp ∞ (α × β)) :
+    nndist x y = nndist x.fst y.fst ⊔ nndist x.snd y.snd :=
+  NNReal.eq <| by
+    push_cast
+    exact prod_dist_eq_sup _ _
+
+variable (p α β)
+
+theorem prod_lipschitzWith_equiv [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    LipschitzWith 1 (WithLp.equiv p (α × β)) :=
+  prod_lipschitzWith_equiv_aux p α β
+
+theorem prod_antilipschitzWith_equiv [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    AntilipschitzWith ((2 : ℝ≥0) ^ (1 / p).toReal) (WithLp.equiv p (α × β)) :=
+  prod_antilipschitzWith_equiv_aux p α β
+
+theorem infty_equiv_isometry [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    Isometry (WithLp.equiv ∞ (α × β)) :=
+  fun x y =>
+  le_antisymm (by simpa only [ENNReal.coe_one, one_mul] using prod_lipschitzWith_equiv ∞ α β x y)
+    (by
+      simpa only [ENNReal.div_top, ENNReal.zero_toReal, NNReal.rpow_zero, ENNReal.coe_one,
+        one_mul] using prod_antilipschitzWith_equiv ∞ α β x y)
+
+/-- Seminormed group instance on the product of two normed groups, using the `L^p`
+norm. -/
+instance instProdSeminormedAddCommGroup [SeminormedAddCommGroup α] [SeminormedAddCommGroup β] :
+    SeminormedAddCommGroup (WithLp p (α × β)) where
+  dist_eq x y := by
+    rcases p.dichotomy with (rfl | h)
+    · simp only [prod_dist_eq_sup, prod_norm_eq_sup, dist_eq_norm]
+      rfl
+    · have : p ≠ ∞ := by
+        intro hp
+        rw [hp, ENNReal.top_toReal] at h
+        linarith
+      simp only [prod_dist_eq_add (zero_lt_one.trans_le h),
+        prod_norm_eq_add (zero_lt_one.trans_le h), dist_eq_norm]
+      rfl
+
+/-- normed group instance on the product of two normed groups, using the `L^p` norm. -/
+instance instProdNormedAddCommGroup [NormedAddCommGroup α] [NormedAddCommGroup β] :
+    NormedAddCommGroup (WithLp p (α × β)) :=
+  { instProdSeminormedAddCommGroup p α β with
+    eq_of_dist_eq_zero := eq_of_dist_eq_zero }
+
+example [NormedAddCommGroup α] [NormedAddCommGroup β] :
+    (instProdNormedAddCommGroup p α β).toMetricSpace.toUniformSpace.toTopologicalSpace =
+    instProdTopologicalSpace p α β :=
+  rfl
+
+example [NormedAddCommGroup α] [NormedAddCommGroup β] :
+    (instProdNormedAddCommGroup p α β).toMetricSpace.toUniformSpace = instProdUniformSpace p α β :=
+  rfl
+
+example [NormedAddCommGroup α] [NormedAddCommGroup β] :
+    (instProdNormedAddCommGroup p α β).toMetricSpace.toBornology = instProdBornology p α β :=
+  rfl
+
+section norm_of
```

Comment:
All the theorems in this section require different typeclass assumptions (you just need a `Norm` to give the statements on norms, and a `dist` to give the statements on `dist`), so it would be better to have something more fine-grained. Also, the `L^2` statements would probably fit better in the `ProdL2` file.